data/src/main/java/me/androidbox/data/local/agenda/AgendaLocalRepositoryImp.kt

```diff
@@ -0,0 +1,124 @@
+package me.androidbox.data.local.agenda
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.supervisorScope
+import me.androidbox.data.local.dao.EventDao
+import me.androidbox.data.local.dao.ReminderDao
+import me.androidbox.data.local.dao.TaskDao
+import me.androidbox.data.mapper.*
+import me.androidbox.domain.agenda.model.Agenda
+import me.androidbox.domain.agenda.model.Event
+import me.androidbox.domain.agenda.model.Reminder
+import me.androidbox.domain.agenda.model.Task
+import me.androidbox.domain.authentication.ResponseState
+import me.androidbox.domain.authentication.remote.AgendaLocalRepository
+import me.androidbox.domain.repository.AgendaRemoteRepository
+import java.time.ZoneId
+import javax.inject.Inject
+
+class AgendaLocalRepositoryImp @Inject constructor(
+    private val eventDao: EventDao,
+    private val taskDao: TaskDao,
+    private val reminderDao: ReminderDao,
+    private val agendaRemoteRepository: AgendaRemoteRepository
+) : AgendaLocalRepository {
+
+    override fun fetchAgenda(startTimeStamp: Long, endTimeStamp: Long): Flow<ResponseState<Agenda>> {
+        return flow<ResponseState<Agenda>> {
+
+            /* Fetch all the events, tasks, and reminders from the DB and return to populate the agenda screen */
+            val events = fetchEventsFromLocal(startTimeStamp, endTimeStamp)
+            val tasks = fetchTasksFromLocal(startTimeStamp, endTimeStamp)
+            val reminders = fetchRemindersFromLocal(startTimeStamp, endTimeStamp)
+
+            /* Emit after fetching from the DB */
+            emit(ResponseState.Success(Agenda(
+                events = events,
+                tasks = tasks,
+                reminders = reminders
+            )))
+
+            /* Fetch all agenda for the current day and time zone */
+            val agenda = agendaRemoteRepository.fetchAgendaForDay(
+                ZoneId.systemDefault(),
+                startTimeStamp * 1000
+            )
+
+            /* Insert events tasks and reminders into the DB */
+            supervisorScope {
+                agenda.events.map { event ->
+                    launch {
+                        eventDao.insertEvent(event.toEventEntity())
+                    }
+                }.forEach { job -> job.join() }
+            }
+
+            supervisorScope {
+                agenda.tasks.map { task ->
+                    launch {
+                        taskDao.insertTask(task.toTaskEntity())
+                    }
+                }.forEach { job -> job.join() }
+            }
+
+            supervisorScope {
+                agenda.reminders.map { reminder ->
+                    launch {
+                        reminderDao.insertReminder(reminder.toReminderEntity())
+                    }
+                }.forEach { job -> job.join() }
+            }
```

Comment:
With this approach you insert all events in parallel, then insert all tasks in parallel and then all reminders. You could do it all in parallel though by first just mapping to the jobs for all types of items and then joining all